pkg/pool/pool.go

```diff
@@ -715,6 +736,73 @@ func (p *pool) getPDDeviation() int {
 	return deviationPrefix * NumIPv4AddrPerPrefix
 }
 
+// calculateSecondaryIPDeviation calculates the deviation required to meet the desired state for secondary IP mode
+// Returns a number of IPv4 addresses by taking into account the MinIPTarget and WarmIPTarget
+func (p *pool) calculateSecondaryIPDeviation() int {
+	numWarmResources := numResourcesFromMap(p.warmResources)
+	numUsedResources := len(p.usedResources)
+	numAssignedResources := numUsedResources + numWarmResources + p.pendingCreate + len(p.coolDownQueue)
+
+	// warm pool is in draining state, set targets to zero
+	if p.warmPoolConfig.DesiredSize == 0 {
+		p.log.V(1).Info("DesiredSize is zero, warmPool is in draining state")
+		p.warmPoolConfig.WarmIPTarget = 0
+		p.warmPoolConfig.MinIPTarget = 0
+		p.warmPoolConfig.WarmPrefixTarget = 0
+	}
+
+	isMinIPTargetInvalid := p.warmPoolConfig.MinIPTarget < 0
+	isWarmIPTargetInvalid := p.warmPoolConfig.WarmIPTarget < 0
+	// Handle scenario where MinIPTarget is configured to negative integer which is invalid
+	if isMinIPTargetInvalid {
+		p.log.V(1).Info(
+			"MinIPTarget value is invalid negative integer, setting MinIPTarget to default",
+			"IPv4DefaultWinMinIPTarget", config.IPv4DefaultWinMinIPTarget,
+		)
+		p.warmPoolConfig.MinIPTarget = config.IPv4DefaultWinMinIPTarget
+	}
+	// Handle scenario where WarmIPTarget is configured to negative integer which is invalid
+	if isWarmIPTargetInvalid {
+		p.log.V(1).Info(
+			"WarmIPTarget value is invalid negative integer, setting warmIPTarget to default",
+			"IPv4DefaultWinWarmIPTarget", config.IPv4DefaultWinWarmIPTarget,
+		)
+		p.warmPoolConfig.WarmIPTarget = config.IPv4DefaultWinWarmIPTarget
+	}
+
+	availableResources := numWarmResources + p.pendingCreate - p.pendingDelete
```

Comment:
`pendingDelete` may not be highly reliable. They could be raced in large scale. By the way, are all those calculations in single goroutine? If the controller support multiple workers, how the atomicity could be guaranteed?

Reply:
> `pendingDelete` may not be highly reliable. They could be raced in large scale.

Noted. I have removed it from the calculation

Reply:
> ...are all those calculations in single goroutine? 

Yes, all calculations are performed within a single goroutine so a race condition isn't expected except with another worker.

> If the controller support multiple workers, how the atomicity could be guaranteed?

I like this kind of PR feedback - its a valid question to ask to help me implement the change correclty. I am going to assume when you say multiple workers you aren't referring to node workers but rather a worker being a controller deployment worker pod.

Short answer is that if the VPC RC changes moves to support multiple workers, then atomicity won't be guaranteed for my changes. In that case, we would need to make some modifications to ensure atomicity so concurrent updates from multiple workers doesn't happen. Some ideas to achieve will likely be some form of distributed locking when making EC2 API requests that mutate the state of the node/ENI.

That being said, as far as I know, **the current implementation uses a single worker scenario**. This is also how prefix delegation mode currently works as it assumes a single worker. The PD mode code will also succumbs to a similar limitation if multiple workers are supported. When using a single leader model, atomicity should not be an issue. 

To provide further confidence on this change is not being negatively impactful, note that in the event of an interrupted worker, the remaining worker can still read the current current state by using the configmap values and ENI state as the source of truth so there wont be an issue with a leader failing mid-update.

Thanks for bringing up the good question. Let me know if you have any other feedback or concerns.

Reply:
Spoke with Hao on this. He wasn't referring to the controller deployments but rather was referring to the struct fields being accessed and in particular the map that is not safe for concurrent use across different goroutines. 

Currently this risk is mitigated by a mutex lock wrapping the entire reconciliation operation but we also discussed that a better implementation might be to wrap map operations with a lock instead rather than making the entire operation lock access to the object. This requires more thought and further refactoring so it has been tabled for a future implementation. 

Resolving conversation.